File: backend/app/schemas/pagination.py
"""Pagination schemas."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class PaginationSchema(BaseModel):
    """Pagination metadata schema."""

    model_config = ConfigDict(from_attributes=True)

    page: int
    per_page: int
    total: int
    pages: int
    has_prev: bool
    has_next: bool
    prev_num: int | None
    next_num: int | None

File: go-backend/internal/service/monitor/pagination.go
package monitor

import "github.com/KanoCifer/kuroome-blog/internal/dto"

// pagination is local to monitor so its endpoint semantics stay independent of
// the other service packages' pagination helpers.
func pagination(page, pageSize, total int) dto.Pagination {
	pages := 0
	if pageSize > 0 {
		pages = (total + pageSize - 1) / pageSize
	}
	prev, next := (*int)(nil), (*int)(nil)
	if page > 1 {
		v := page - 1
		prev = &v
	}
	if page < pages {
		v := page + 1
		next = &v
	}
	return dto.Pagination{
		Page: page, PerPage: pageSize, Total: total, Pages: pages,
		HasPrev: page > 1, HasNext: page < pages, PrevNum: prev, NextNum: next,
	}
}
